### ISSUE TITLE
Understanding the loss implementation

### ISSUE THREAD

**Issue by djo-koconi** [anicolson/DeepXi#33]
Hi @anicolson .

Please help me understand how binary crossentropy loss is used to train Deep Xi. 
I understand the idea and the implementation, but can not fully comprehend the theory behind the loss.

As binary crossentropy implies a kind of classification task, how we can eventually get the right prediction?
I do not understand how is the mapping done between the input features (mag specs) and the targets (SNRs). 

Also, I have red the original paper and couldn't find an answer there.

**Comment by anicolson** [comment]
Hi,

It is important that you understand the mapping function (the CDF of each frequency bin), as it allows the target values to be in the set (0,1).

As just stated, each target component is in the set (0,1). Also, each output node of the network can take on a value the set (0,1), as a sigmoid activation function is used. You can think of each output node representing the probability that the frequency bin is completely speech (i.e. 1 means completely speech, 0 means no speech, and somewhere in between indicates that a proportion is speech). 

Binary cross-entropy (BCE), as opposed to categorical cross-entropy (CCE), can be used as the loss function for this (CCE is used for a categorical distribution over the output nodes, which we are not interested in). With BCE, the target values must be in the set (0,1) (they don't have to be strictly 0 and 1). BCE just measures the difference between the target probability and predicted probability. For pure classification, the target probability is either 0 or 1. Here, the target probability is between 0 and 1 (essentially an output node is learning the CDF of speech in a frequency bin).

This was found to be a better alternative to having the *a priori* SNR as the target and using mean squared error (MSE).

Hope this helps, 

Aaron.

**Comment by djo-koconi** [comment]
Thank you so much @anicolson !
Its much more clear now.

If I can try to rephrase, we are basically mapping the probability that spectral component is containing noise, to the input feature (spectral component distribution) itself. 
Therefore, we use BCE to measure the difference between the predicted probability that spectral component is containing noise and the target probability, which we already know.
 
There are both in range of {0,1} which describe the amount of noise (or speech) that freq bin contains.

**Comment by anicolson** [comment]
As it is the CDF of instantaneous a priori SNR, we are mapping the probability that a spectral component contains completely speech (not noise) as 1 indicates that the component is completely speech. If it was flipped, i.e. noise-to-signal ratio, then it would be the probability that the component is completely noise.

Hope this helps,

Aaron

**Comment by djo-koconi** [comment]
Yes sure, thanks again!

By the  way, hope I'm not bothering you too much :) 
Any specific reason for choosing Xi in naming your model? Why that specific Greek letter?

**Comment by anicolson** [comment]
The a priori SNR is represented by xi in the literature.

**Comment by djo-koconi** [comment]
Aha, ok, thanks.
Sorry for being ignorant of the fact :)

**Comment by anicolson** [comment]
Happy to help.

**Comment by Shipley1105** [comment]
the target value of `BinaryCrossentropy` can be only choosed 0 or 1, as far as i am concerned target value(cdf) range from 0 to 1.
If using `BinaryCrossentropy` DeepXi can be regarded as a frequency-bin level VAD, and the training target only choosed 0 or 1, am I right?
I stilll don't understand why  `BinaryCrossentropy` is choosed. @anicolson Could you give me some explanation?

**Comment by anicolson** [comment]
Hi Shipley,

BCE is chosen over categorical cross entropy as the latter expects the target to be a one-hot vector (i.e., a model with a softmax output layer activation function). Cross entropy is a measure of difference between distributions. The documentation is a bit misleading; the values of the target can be between 1 and 0. Here is the documentation:

https://www.tensorflow.org/api_docs/python/tf/keras/losses/BinaryCrossentropy

It has some exampes:
![image](https://user-images.githubusercontent.com/26111230/216200732-4bb2c57a-3bff-4114-883d-8270c3ee8fad.png)

Try these examples with y_true values between 0 and 1 and you can see how the cross entropy changes. 

Targets tend to be 0 or 1 for classification, which we dont want. We want to estimate the cumulative probability, so the cross entropy in our case lets us know the difference between the estimated and target cumulative probability.

Hope this helps!
Aaron
